Serialize repair with tablet migration

tgrabiec opened the pull request:
We want to exclude repair with tablet migrations to avoid races
between repair reads and writes with replica movement. Repair is not
prepared to handle topology transitions in the middle.

One reason why it's not safe is that repair may successfully write to
a leaving replica post streaming phase and consider all replicas to be
repaired, but in fact they are not, the new replica would not be
repaired.

Other kinds of races could result in repair failures. If repair writes
to a leaving replica which was already cleaned up, such writes will
fail, causing repair to fail.

Excluding works by keeping effective_replication_map_ptr in a version
which doesn't have table's tablets in transitions. That prevents later
transitions from starting because topology coordinator's barrier will
wait for that erm before moving to a stage later than
allow_write_both_read_old, so before any requests start using the new
topology. Also, if transitions are already running, repair waits for
them to finish.

A blocked tablet migration (e.g. due to down node) will block repair,
whereas before it would fail. Once admin resolves the cause of blocked migration,
repair will continue.  

- ** Backport reason **
No need for backport because tablets are experimental in earlier releases

denesb commented:
@tgrabiec doesn't this fix https://github.com/scylladb/scylladb/issues/17658?

raphaelsc commented on repair/repair.cc:
```diff
@@ -2102,8 +2102,19 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
+            rlogger.info("repair[{}] Table {}.{} has tablet transitions, waiting for topology to quiesce", rid.uuid(), keyspace_name, table_name);
+            erm = nullptr;
+            co_await _ss.local().await_topology_quiesced();
+            rlogger.info("repair[{}] Topology quiesced", rid.uuid());
```
should await_topology_quiesced() return group0_guard? maybe not for this usage, but for others that might yield and want to read group0 state.

tgrabiec replied:
There's no need, the caller can also obtain the guard. If topology is not quiesced then, he can retry. That's what await_topology_quiesced() would have to do internally anyway to return the guard.

asias replied:
After the while (true) loop exits, what does it guarantee exactly?

Will holding a copy of the erm prevent any future topology operations / tablet migration? If so, how does it prevent the operations under the hood?

tgrabiec replied:
It guarantees that erm doesn't have tablets in transition.

Holding to that erm prevents transitions from starting, because topology coordinator waits for old erms to be released when it executes global token metadata barrier.

asias replied:
> It guarantees that erm doesn't have tablets in transition.

Only for the table in question which the erm belongs to, or for all tables?

Or even stronger, that no tablets (from any tables) in transitions and no topology changes cluster wide.  Since await_topology_quiesced waits for topology_state_machine to be not busy, it means the cluster is not in any of the states (tablet_draining, tablet_migration etc.).

```
future<> storage_service::await_topology_quiesced() {
    auto holder = _async_gate.hold();

    if (this_shard_id() != 0) {
        // group0 is only set on shard 0.
        co_await container().invoke_on(0, [&] (auto& ss) {
            return ss.await_topology_quiesced();
        });
        co_return;
    }

    group0_guard guard = co_await _group0->client().start_operation(&_group0_as, raft_timeout{});

    while (_topology_state_machine._topology.is_busy()) {
        rtlogger.debug("await_topology_quiesced(): topology is busy");
        release_guard(std::move(guard));
        co_await _topology_state_machine.event.wait();
        guard = co_await _group0->client().start_operation(&_group0_as, raft_timeout{});
    }
}

bool topology::is_busy() const {
    return tstate.has_value();
}   

struct topology {
    enum class transition_state: uint16_t {
        join_group0,
        commit_cdc_generation,
        tablet_draining,
        write_both_read_old,
        write_both_read_new,
        tablet_migration,
        tablet_split_finalization,
        left_token_ring,
        rollback_to_normal,
    };

    std::optional<transition_state> tstate;

}
```

> Holding to that erm prevents transitions from starting, because topology coordinator waits for old erms to be released when it executes global token metadata barrier.

Including any tablet migration and adding/removing node as well?

tgrabiec replied:
> Only for the table in question which the erm belongs to, or for all tables?

Only for the table, because that's what we check for.

tgrabiec replied:
> Including any tablet migration and adding/removing node as well?

Yes.

tgrabiec replied:
> Since await_topology_quiesced waits for topology_state_machine to be not busy, it means the cluster is not in any of the states (tablet_draining, tablet_migration etc.).

Yes, it waits for !busy, but this can change before we obtain the erm, so it's not guaranteed. What is guaranteed is what we check about the erm state.

gleb-cloudius replied:
So given that why check `busy` at all? Isn't it enough to do:

```
while (auto erm = t->get_effective_replication_map(); erm.has_transitions()) {
    erm.reset();
    topology.even.wait();
}
```

tgrabiec replied:
I think it's not that simple because topology is only valid on shard 0, so we cannot atomically check erm (on non-0 shard) and call wait(). We would be prone to missed notification if we check erm across shards, if topology becomes non-busy before we wait(). If we check erm on shard-0, then we can't use it on the original shard. We'd have to check the erm on both shard-0 and here.

gleb-cloudius replied:
Will this work:

```
while (auto erm = t->get_effective_replication_map(); erm.has_transitions()) {
    erm.reset();
    invoken_on(0, [] () {
        if (t->get_effective_replication_map().has_transitions()) {
            topology.event.wait();
        }
   }
}
```

tgrabiec replied:
No because "t" cannot be accessed across shards. We'd have to re-lookup it. 

Calling quiesce() is simpler.

gleb-cloudius replied:
> No because "t" cannot be accessed across shards. We'd have to re-lookup it.
> 

Well, of course I did not mean the code to be 100% correct. erm does not have  `has_transitions` either.

> Calling quiesce() is simpler.

May be, but not much TBH. I just dislike the function since we still argue about what it actually guaranties and would gladly get rid of it.

tgrabiec replied:
> May be, but not much TBH. I just dislike the function since we still argue about what it actually guaranties and would gladly get rid of it.

I can rename it to `await_not_busy()` to avoid confusion about semantics.

gleb-cloudius replied:
It is better, but if this is the only change no need to delay the PR just for that.

Also @bhalevy here https://github.com/scylladb/scylladb/pull/18641#discussion_r1624511471 asked if we can wait for a specific tablet which I think my proposal would do.

gleb-cloudius replied:
> > Also @bhalevy here [#18641 (comment)](https://github.com/scylladb/scylladb/pull/18641#discussion_r1624511471) asked if we can wait for a specific tablet
> 
> Specific table, not tablet. That latter is too complicated at the moment. I'm not sure about the former.

Yes, I meant table.

tgrabiec replied:
> Also @bhalevy here https://github.com/scylladb/scylladb/pull/18641#discussion_r1624511471 asked if we can wait for a specific tablet which I think my proposal would do.

Yes, but I don't think we need to optimize for this. I'd rather invest in implementing to proper solution, which is to schedule repair from tablet scheduler.

asias commented on repair/row_level.cc:
```diff
@@ -3178,7 +3178,8 @@ class row_level_repair_gossip_helper : public gms::i_endpoint_state_change_subsc
 };
 
-repair_service::repair_service(distributed<gms::gossiper>& gossiper,
+repair_service::repair_service(distributed<service::storage_service>& ss,
```
We already pass repair_service to repair_service.  This makes circular dependency.  

```
    storage_service(abort_source& as, distributed<replica::database>& db,
...
        sharded<repair_service>& repair,
```

asias commented on repair/repair.cc:
```diff
@@ -2102,8 +2102,19 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
+            }
+            rlogger.info("repair[{}] Table {}.{} has tablet transitions, waiting for topology to quiesce", rid.uuid(), keyspace_name, table_name);
+            erm = nullptr;
+            co_await _ss.local().await_topology_quiesced();
```
We need to find a way to avoid using _ss service inside repair_service.

mykaul commented:
> DO NOT MERGE: Depends on #18026 (review last 4 commits).

I think this initial comment at least can be removed, as #18026 is merged already.

tgrabiec commented:
In V2:
 - avoid cyclic dependency between repair_service and storage_service by extracting topology_state_machine
 - rebased

kbr-scylla commented:
@asias you had comments. Do you approve now?

asias commented:
> @asias you had comments. Do you approve now?

Yes. Looks good to me. Thanks.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6178,14 +6178,8 @@ future<> storage_service::await_topology_quiesced() {
         co_return;
     }
 
```
> and if it's false it means the topology is no longer busy.

I already complained once that it is not cleat to me what this function is actually tries to do, but the definition above is incorrect. If it is false it means that the topology was not busy at least once after the function was called. What is this useful for I cannot imagine.

tgrabiec replied:
> the definition above is incorrect. If it is false it means that the topology was not busy at least once after the function was called.

Of course the condition can change again as soon as we notice it, but you can say that about every function which waits for a predicate. This is still useful because if the topology became not busy at some point it means that the backlog created by earlier events was cleared at some point. For example, after replacing a node, replicas were rebuilt.

gleb-cloudius replied:
> > the definition above is incorrect. If it is false it means that the topology was not busy at least once after the function was called.
> 
> Of course the condition can change again as soon as we notice it, but you can say that about every function which waits for a predicate. 

There are functions that take a lock and guaranty that a predicate is true after returning. There are cases where there is no concurrency and then the condition is true until next  yield. Those are not the cases here of course.

> This is still useful because if the topology became not busy at some point it means that the backlog created by earlier events was cleared at some point. For example, after replacing a node, replicas were rebuilt.

True, but you need to make sure that the request started to execute. I am not sure we can guaranty that the coordinator never becomes non busy until all request are processed. It may become non busy between requests I think, but if not is it an implementation detail we want to rely on?

kostja replied:
It's unclear why repair will ever execute on a busy system. Individual tablet movements don't block each other, while a single tablet movement block repair. The waiting is done without any kind of a pending lock on the topology that would prevent new tablet movements from starting. 

It's important to achieve some kind of FIFO between tablet movements and repair. I guess it's in the plan, so this comment is not a review request, it's a follow up request and question.

tgrabiec replied:
> It's unclear why repair will ever execute on a busy system. Individual tablet movements don't block each other, while a single tablet movement blocks repair. The waiting is done without any kind of a pending lock on the topology that would prevent new tablet movements from starting.
> 
> It's important to achieve some kind of FIFO between tablet movements and repair. I guess it's in the plan, so this comment is not a review request, it's a follow up request and question.

Yes, there is no preemption, so rebalancing will block repair. This will be addressed when we implement the original design to track repair requests in group0 and schedule tablet repair from tablet schduler.

tgrabiec replied:
> True, but you need to make sure that the request started to execute. I am not sure we can guaranty that the coordinator never becomes non busy until all request are processed. It may become non busy between requests I think, but if not is it an implementation detail we want to rely on?

Note that this patch doesn't rely on this. But I think we should guarantee it one way or the other. We could invoke the load-balancer to check if !busy topology has backlog. But it seems expensive, so guaranteeing in the coordinator that !busy topology implies no load-balancer backlog may be better, and not that hard (just call the load balancer on the exit edges).

gleb-cloudius replied:
How expensive is to check the backlog? If we want some kind of guaranty from the coordinator we need to document it and make it somehow hard to break without noticing.

tgrabiec replied:
> How expensive is to check the backlog? 

It's O(node_count + tablet_replica_count) if no balancing is needed. Not the end of the world, depending on how frequently we check busy(). And assuming that load balancer will stay stateless, so that it can be called not only by the coordinator but from any node and shard.

gleb-cloudius commented on main.cc:
```diff
@@ -1442,14 +1442,21 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
                 qp.invoke_on_all(&cql3::query_processor::stop_remote).get();
             });
 
+            sharded<service::topology_state_machine> tsm;
```
Should it be sharded at all? It is valid on shard0 only.

tgrabiec replied:
Good question. But it was sharded before, every storage_service instance had it. If we don't make it sharded, we would have to make it optional and disengaged on non-zero shards. It would be a bit more complicated to propagate to sharded components.

gleb-cloudius replied:
Yes. Lets leave it to a follow up.

tgrabiec commented:
> Is this for general repair (vnode based too), or only tablet repair?

Only tablet repair.

tgrabiec commented on service/topology_state_machine.cc:
```diff
@@ -234,6 +234,12 @@ topology::upgrade_state_type upgrade_state_from_string(const sstring& s) {
 
+future<> topology_state_machine::await_quiesced() {
+    while (_topology.is_busy()) {
+        co_await event.wait();
```
Isn't event broken on abort?

tgrabiec replied:
@gleb-cloudius ?

gleb-cloudius replied:
It is:

```
future<> storage_service::wait_for_group0_stop() {
    _group0_as.request_abort();
    _topology_state_machine.event.broken(make_exception_ptr(abort_requested_exception()));
    co_await when_all(std::move(_raft_state_monitor), std::move(_sstable_cleanup_fiber), std::move(_upgrade_to_topology_coordinator_fiber));
}

```

tgrabiec commented:
In V2.1:
  - Fixed "Fixes ..." tag to reference the issues
  - Rebased

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9401/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9401/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_experimental_raft/test_tablets
#### Failed Tests (1/5400):
- [test_tablet_missing_data_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9401/testReport/junit/%28root%29/test_tablets/test_tablet_missing_data_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_missing_data_repair)
#### Build Details:
- Duration: 4 hr 59 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9410/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9410/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_experimental_raft/test_tablets
#### Failed Tests (2/5400):
- [test_tablet_missing_data_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9410/testReport/junit/%28root%29/test_tablets/test_tablet_missing_data_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_missing_data_repair)
- [test_tablet_missing_data_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9410/testReport/junit/%28root%29/test_tablets/test_tablet_missing_data_repair_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_missing_data_repair)
#### Build Details:
- Duration: 4 hr 56 min
- Builder: spider2.cloudius-systems.com

bhalevy commented on service/storage_service.cc:
```diff
@@ -6200,14 +6200,8 @@ future<> storage_service::await_topology_quiesced() {
-        co_await _topology_state_machine.event.wait();
-        guard = co_await _group0->client().start_operation(&_group0_as, raft_timeout{});
-    }
+    co_await _group0->group0_server().read_barrier(&_group0_as);
```
So the `read_barrier` replaces the guard we get from `start_operation` in the replaced code?

bhalevy commented on repair/repair.cc:
```diff
@@ -2106,8 +2106,21 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
+            rlogger.info("repair[{}] Table {}.{} has tablet transitions, waiting for topology to quiesce", rid.uuid(), keyspace_name, table_name);
+            erm = nullptr;
+            co_await container().invoke_on(0, [] (repair_service& rs) {
+                return rs._tsm.local().await_quiesced();
```
> We do not have a task covering the repair preparation part

Maybe we should have one if quiescing topology may block for a long time and the user wants to abort repair.

@tgrabiec could repair block behind decommission or rf change, that might take hours?  or would it wait just behind one batch, taking, say, minutes?

tgrabiec replied:
> Also, I'm concerned that await_quiesced would continue poll without preventing new topology changes to start.

Yes, but to fix that, we'd have to implement the original design in which repairs requests are stored in group0 and tablet scheduler can see them as preemption request.

> @tgrabiec could repair block behind decommission or rf change, that might take hours? or would it wait just behind one batch, taking, say, minutes?

It will block behind whole decommission.

bhalevy replied:
> > Also, I'm concerned that await_quiesced would continue poll without preventing new topology changes to start.
> 
> Yes, but to fix that, we'd have to implement the original design in which repairs requests are stored in group0 and tablet scheduler can see them as preemption request.

Okay.  So it's work for a followup.

> > @tgrabiec could repair block behind decommission or rf change, that might take hours? or would it wait just behind one batch, taking, say, minutes?
> 
> It will block behind whole decommission.

I see.  As follow up, it would be good to make it by an async, abortable task.
Then, assuming repairs will be scheduled in single-tablet resolution, they could happen before migration for decommission, or after, or if we go very crazy, we can do repair-based migration instead of regular migration (but I don't expect that to be done any time soon, or ever).

tgrabiec replied:
> Also, can we wait until the specific table is quiesced rather than waiting for the whole topology state machine to stop being busy?

We could, but it's an optimization I'm not sure is worth it. If there is an ongoing topology operation, it probably affects many tables, so it may not help much in practice.

mykaul commented:
@bhalevy , @tgrabiec - what's the status of this PR?

tgrabiec commented:
In V3:
  - renamed topology_state_machine::await_quiesced() to await_not_busy()
  - Added a fix for CI stability: [test: pylib: Do not block async reactor while removing directories](https://github.com/scylladb/scylladb/pull/18641/commits/5ca54a6e8818b72a0634d0123cc110529d994354)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9488/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9488/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_experimental_raft/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9488/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1908/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9933/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9488/)
#### Build Details:
- Duration: 7 hr 4 min
- Builder: spider5.cloudius-systems.com

tgrabiec commented on test/pylib/scylla_cluster.py:
```diff
@@ -353,7 +362,7 @@ async def install(self) -> None:
 
         # Cleanup any remains of the previously running server in this path
-        shutil.rmtree(self.workdir, ignore_errors=True)
+        await async_rmtree(self.workdir, ignore_errors=True)
```
@kbr-scylla @kostja Please review.

bhalevy replied:
should we submit this separately?
we may want to backport this particular patch to older releases as well to reduce flakiness there

kbr-scylla replied:
It's enough for it to be separate commit to backport

kbr-scylla replied:
@xtrey please review this commit (https://github.com/scylladb/scylladb/commit/5ca54a6e8818b72a0634d0123cc110529d994354)

kbr-scylla commented:
Qd
